model/host/host.go

```diff
@@ -746,6 +741,45 @@ func (h *Host) UpdateProvisioningToRunning() error {
 	return nil
 }
 
+// SetNeedsJasperRestart sets this host as needing to have its Jasper service
+// restarted as long as the host does not already need a different
+// reprovisioning change. If the host is ready to reprovision now (i.e. no agent
+// monitor is running), it is put in the reprovisioning state.
+func (h *Host) SetNeedsJasperRestart() error {
+	if err := h.setAwaitingJasperRestart(); err != nil {
+		return err
+	}
+	if h.StartedBy == evergreen.User && !h.NeedsNewAgentMonitor {
+		return nil
+	}
+	return h.MarkAsReprovisioning()
```

Comment:
are we worried about the race/non-transactional nature of these operations?

Reply:
Not really. I did a sanity check for the reprovisioning field is the host field in-memory so if it updates and the field has been cleared, then MarkAsReprovisioning will just error and not set the state to reprovisioning.